cas-server-core-tickets/src/main/java/org/jasig/cas/ticket/OAuthCodeImpl.java

```diff
@@ -0,0 +1,118 @@
+package org.jasig.cas.ticket;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.jasig.cas.authentication.Authentication;
+import org.jasig.cas.authentication.principal.Service;
+import org.jasig.cas.ticket.proxy.ProxyGrantingTicket;
+import org.springframework.util.Assert;
+
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorColumn;
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+import javax.persistence.Lob;
+import javax.persistence.Table;
+import javax.validation.constraints.NotNull;
+
+/**
+ * An OAuth code implementation.
+ *
+ * @author Jerome Leleu
+ * @since 4.3.0
+ */
+@Entity
+@Table(name="OAUTH_TOKENS")
+@DiscriminatorColumn(name="TYPE")
+@DiscriminatorValue(OAuthCode.PREFIX)
+public class OAuthCodeImpl extends AbstractTicket implements OAuthCode {
```

Comment:
That was my original idea. But in that case, OAuth tickets will be available through the `ServiceTicketDelegator` which misses the `getAuthentication` method and thus needs to make public its `getTicket` method: ugglier from my point of view compared to a clean hierarchy as I propose.